### PR TITLE
Assignment1

### PR DESCRIPTION
**Pull request by etheodoridou**
Final Assignment 1 submission

### CODE REVIEW

File: Exercise1.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python
+
+
+def is_permutation(string1, string2) :
```
Comment (flerdacodeu) [inline Exercise1.py]: Add a docstring comment to your function describing what it is meant to do, what the arguments are, and what it is meant to return, since this is the main function in your file.

File: Exercise1.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	# check if any of the strings is None
```
Comment (flerdacodeu) [inline Exercise1.py]: Avoid comments that are sort of obvious: in this case the comment is saying exactly what the code is saying, in as simple terms. Save comments for explaining complicated code (if it cannot be made simpler) or to explain why you are doing something, which the code itself cannot explain.

File: Exercise1.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	if string1 is None or string2 is None:
+		return False
+	# change strings inputted to dictionaries
+	string1_dict = string_to_dict(string1)
+	string2_dict = string_to_dict(string2)
+	
+	# check if the two dictionaries are equal
+	result = dicts_are_equal(string1_dict, string2_dict)
+	
+	return result
```
Comment (flerdacodeu) [inline Exercise1.py]: You could avoid the `result` directory are return the result directly.

File: Exercise1.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	
+def string_to_dict(input_string) :
+	# make sure all characters are lower case
+	lower_case_string = input_string.lower()
+	
+	# create a dictionary where you map each character with a counter of
+	# how many instances of a character exist
+	# if the strings can only include characters from the English alphabet:
```
Comment (flerdacodeu) [inline Exercise1.py]: I am not sure I understand what this comment is meant to signify.
Is there something we need to do here? Or is the code complete without this (in which case we can delete this comment).

File: Exercise1.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	# initialise the dictionary at this stage with all the letters of the alphabet
+	# and a starting counter value of 0
+	dict = {}
+	
+	# add the characters to the dictionary
+	for character in lower_case_string:
+		if character in dict:
```
Comment (flerdacodeu) [inline Exercise1.py]: You might want to use the `get` function defined on `dict` to simplify this code.

File: Exercise1.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			value = dict[character]
+			dict[character] = value + 1
+		else:
+			dict[character] = 1
+			
+	# for the case where only the English alphabet is allowed
```
Comment (flerdacodeu) [inline Exercise1.py]: I think I understand your above comment: you are suggesting an alternative solution.
The way I would do that would be to write a separate function:

```python
def string_to_dict_english(input_string):
  [...]
```

so that you can define this function (and test it!).

Alternatively, you can just remove the commented out code.
Commented out code tends to not be useful in the long term.

File: Exercise1.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	# for character in lower_case_string:
+	# 	if character in dict:
+	# 		value = dict[character]
+	# 		dict[character] = value + 1
+	# 	else:
+	# 		return None
+	
+	return dict
+
+def dicts_are_equal(dict1, dict2) :
```
Comment (flerdacodeu) [inline Exercise1.py]: You can just compare two dictionary in Python :)

File: Exercise1.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	# and the counter values are equal
+	for key in dict1:
+		if key in dict2 and dict2[key] == dict1[key]:
+			continue
+		else:
+			return False
+	return True
+
+def main() :
+	string1 = raw_input('Please insert the first string.')
```
Comment (flerdacodeu) [inline Exercise1.py]: Small suggestion: add a space after the end of the string. Also, make both questions or use ':' at the end.

File: Exercise1_Test.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env python
+
+import Exercise1
+import unittest
+
+class Exercise1Test(unittest.TestCase):
+
+  def testBaseCases(self):
+    self.assertEquals(True, Exercise1.is_permutation("", ""))
```
Comment (flerdacodeu) [inline Exercise1_Test.py]: Use can use:

```python
  self.assertTrue([...])
  self.assertFalse([...])
```

File: Exercise1_Test.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    self.assertEquals(False, Exercise1.is_permutation(None, None))
+
+  def testKnownCases(self):
```
Comment (flerdacodeu) [inline Exercise1_Test.py]: Can you think of other corner cases?

File: KthToLastElement_Test.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
+
+from SimplyLinkedList import SimplyLinkedList
```
Comment (flerdacodeu) [inline KthToLastElement_Test.py]: This does look a bit confusing.
How about calling the module `linked_list.py` instead?

File: Link.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python
+
+class Link(object) :
+	
+	def __init__(self, value, next_node=None):
```
Comment (flerdacodeu) [inline Link.py]: Since you call the argument `next_node` maybe `Node` is a better name for this class.

File: SimplyLinkedList.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
+
+from Link import Link
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: The `Link` is an implementation detail and can be kept into the same file as the where `SimplyLinkedList` is.

File: SimplyLinkedList.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+	def remove(self, data):
+		# get the current link
+		current = self.head
+		# get the previous link
+		previous = None
+		# save whether the falue has been found
+		found = False
+		
+		# while the link is not None and the value has not been found
+		while current and found is False:
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: Instead of `found is False` you can use `not Found`, which is easier to understand.

File: SimplyLinkedList.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+			# if data is found update boolean
+			if current.get_value() == data:
+				found = True
+			# else move on to the next link
+			else:
+				previous = current
+				current = current.get_next()
+		
+		# if the current link is None the data has not been found
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: Why not use the `found` variable instead? It seems more obvious.
Alternatively, you can get rid of the `found` variable and use `break` in your loop.

File: SimplyLinkedList.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+			raise ValueError("Data not in list")
+		
+		# replace the current item with the item after it
+		# removing it hence from your list
+		previous.set_next(current.get_next())
+	
+	def kth_to_last_element(self, head, k):
+		# if the index provided is higher than the size of the array
+		# alert the user
+		if k > self.size : raise ValueError("Index not in list")
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: `self.size` is actually a method.
It does not make sense to compare it with an integer.
You should tests to your code to check this case.

File: SimplyLinkedList.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+		
+		# when you reach the end of the list return
+		if head is None: return 0
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: Can you think of a better value to return than 0 here?

File: SimplyLinkedList.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+		
+		# update the index of the current link in the list
+		index = self.kth_to_last_element(head.get_next(), k) + 1
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: This implementation is rather inefficient: you call `size()`, which is `O(n)` at the beginning of the call and then call this recursively up to `n` times, resulting in this being `O(n^2)`. This function does really need to be recursive and you do not need to check `size()` explicitly.

File: SimplyLinkedList.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+		
+		# if the current link is None the data has not been found
+		if current is None:
+			raise ValueError("Data not in list")
+		
+		# replace the current item with the item after it
+		# removing it hence from your list
+		previous.set_next(current.get_next())
+	
+	def kth_to_last_element(self, head, k):
```
Comment (flerdacodeu) [inline SimplyLinkedList.py]: Add a docstring for this function, stating what the function is meant to do, what are the arguments, and what it returns.

*NOTE: Your implementation returns the `index` of the k-th to last element, rather than the value.*

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
I added a few comments.

Main high-level comments are:
* Your k-th to last element function returns the index of that element rather than its value.
* You are missing tests for your assignment.

File: Exercise1.py (hunk 7 above)
Comment (etheodoridou) [reply Exercise1.py]: Do you mean to compare them by using "=="?

File: Exercise1.py [hunk 20]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
+
+"""
```
Comment (flerdacodeu) [inline Exercise1.py]: The docstring should be inside the function itself.

See https://www.python.org/dev/peps/pep-0257/ for details.

If should probably look like:

```python
  def is_permutation(string1, string2):
    """Determines if two strings are permutations of each other.

    Args:
      string1: the first string to compare
      string2: the second string to compare

    Returns:
       True if the two strings are permutations of each other, otherwise False.
    """
```

File: Exercise1.py [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+			dict[character] = dict.get(character) + 1
+		else:
+			dict[character] = 1
+
+	return dict
+
+def string_to_dict_english(input_string):
+    lower_case_string = input_string.lower()	
+	
+	dict = {}
```
Comment (flerdacodeu) [inline Exercise1.py]: Indentation here seems wrong.

File: Exercise1.py [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+	
+	for alphabet in range('a','z'):
+		dict[character] = 0
+	
+	for character in lower_case_string:
+		if character in dict:		
+			dict[character] = dict.get(character) + 1
```
Comment (flerdacodeu) [inline Exercise1.py]: In this case, you can just use `+=`.

File: Exercise1.py (hunk 7 above)
Comment (flerdacodeu) [reply Exercise1.py]: Yes, that's right! In Python the `==` operator over dictionaries does exactly what you need here.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]

File: Exercise1.py (hunk 1 above)
Comment (etheodoridou) [reply Exercise1.py]: Done!

File: Exercise1.py (hunk 2 above)
Comment (etheodoridou) [reply Exercise1.py]: Done! Deleted all useless comments

File: Exercise1.py (hunk 3 above)
Comment (etheodoridou) [reply Exercise1.py]: Done!

File: Exercise1.py (hunk 5 above)
Comment (etheodoridou) [reply Exercise1.py]: I am not really sure how to use get in lines 29/30 but I did use it in line 31 to avoid saving the counter in a value variable first.

File: Exercise1.py (hunk 6 above)
Comment (etheodoridou) [reply Exercise1.py]: I decided to add the recommended method to my implementation.

File: Exercise1.py (hunk 7 above)
Comment (etheodoridou) [reply Exercise1.py]: Okay I have used that now!

File: Exercise1_Test.py (hunk 9 above)
Comment (etheodoridou) [reply Exercise1_Test.py]: Done!

File: Exercise1_Test.py (hunk 10 above)
Comment (etheodoridou) [reply Exercise1_Test.py]: I have added a few more tests for the method that handles only English and some more tests for the base case. I also wanted to test how the maximum int value will be dealt with, but my test never terminates. See below:
`def testErrorCases(self):`
`    value = ""`
`    for counter in range(0, sys.maxsize) :`
`        value += "a"`
`    self.assertTrue(Exercise1.is_permutation(value, value))`

File: Exercise1.py (hunk 22 above)
Comment (etheodoridou) [reply Exercise1.py]: Done!

File: Exercise1.py (hunk 21 above)
Comment (etheodoridou) [reply Exercise1.py]: Indentation fixed!

File: Exercise1.py (hunk 20 above)
Comment (etheodoridou) [reply Exercise1.py]: Noted. I have changed the comment format :)

File: Exercise1_Test.py (hunk 10 above)
Comment (flerdacodeu) [reply Exercise1_Test.py]: Since `sys.maxsize` is very very large (`9223372036854775807` for my installation of Python) a string with that many characters would require 9 exabytes of memory (9 billion gigabytes). The tests are probably taking a long time as they require more and more memory (and you are actually each time allocating a new strings with yet one more character in it). So this is not really feasible to test.

I would give up on such a test: you can use a long string, but `sys.maxsize` is just too much.

For the case of a long string, you might also want to consider the following:

```python
>>> "a" * 10 == "aaaaaaaaaa"
True
```

File: Exercise1.py (hunk 20 above)
Comment (flerdacodeu) [reply Exercise1.py]: I think you might have forgotten to upload the latest copy of the code?

**Comment by flerdacodeu** [comment]
Hello! I think the code for the updates was not uploaded.
When you have a chance, can you upload it?
So that we can close this pull request and merge the changes into master.

File: SimplyLinkedList.py (hunk 13 above)
Comment (etheodoridou) [reply SimplyLinkedList.py]: Could you explain this a bit further? Since I come from a Java background, I am used to having separate files for each class/object.

File: KthToLastElement_Test.py (hunk 11 above)
Comment (etheodoridou) [reply KthToLastElement_Test.py]: Done!

File: Link.py (hunk 12 above)
Comment (etheodoridou) [reply Link.py]: I decided to change all naming to Link

File: SimplyLinkedList.py (hunk 14 above)
Comment (etheodoridou) [reply SimplyLinkedList.py]: Done!

File: SimplyLinkedList.py (hunk 15 above)
Comment (etheodoridou) [reply SimplyLinkedList.py]: Fixed!

File: SimplyLinkedList.py (hunk 19 above)
Comment (etheodoridou) [reply SimplyLinkedList.py]: Added! 

I changed the implementation to the one given as a solution as I couldn't change my initial implementation to return the value rather than the index and I also added unittests.

File: SimplyLinkedList.py (hunk 17 above)
Comment (etheodoridou) [reply SimplyLinkedList.py]: I raised a ValueError instead when the index given is out of bounds. Is that a better approach?

**Comment by etheodoridou** [comment]
Thank you for your review! Apologies for the delay. I have now updated all my files. Could you have another look please? :)

File: SimplyLinkedList.py (hunk 17 above)
Comment (flerdacodeu) [reply SimplyLinkedList.py]: Yes, raising an exception is a good way to handle this.

File: SimplyLinkedList.py (hunk 13 above)
Comment (flerdacodeu) [reply SimplyLinkedList.py]: Of course!

In Java, each class is in a file by the same name.

In Python, each file is called a module and can contain multiple classes or functions.
That is because the file itself defines the name of the module. This allows us, among other things, to hide within the module all the implementation details (e.g., the `Link` class here that is not needed by someone using the `SimplyLinkedList`.

You can find of a Python file a little bit like a package in Java, where you can have multiple classes with it.

Usually it is a good rule to hide any implementation details. In Java you could do this with a private nested class (which is accessible only from the class in which it is defined) or a package-private class (which is accessible only from classes in the same package).

File: assignment1/Exercise1.py [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+def string_to_dict(input_string) :
+    lower_case_string = input_string.lower()
+    
+    # create a dictionary where you map each character with a counter of
+    # how many instances of a character exist
+    dictionary = {}
+    
+    for character in lower_case_string:
+        if character in dictionary:
+            dictionary[character] += 1
```
Comment (flerdacodeu) [inline assignment1/Exercise1.py]: You can use the `get` function to remove the if statement.
For example:
```python
  dictionary.get('A', 42)
```
will return the value stored in the dictionary for the string `'A'` if present, or `42` if not present.
Of course, `'A'` and `42` are just examples.

File: assignment1/Exercise1.py [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    dictionary = {}
+    
+    for alphabet in ascii_lowercase:
+        dictionary[alphabet] = 0
+    
+    for character in lower_case_string:
+        if character in dictionary:        
+            dictionary[character] = dictionary.get(character) + 1 
+        else:       
+            return {}
```
Comment (flerdacodeu) [inline assignment1/Exercise1.py]: This is probably better handled as an exception as well.

File: assignment1/KthToLastElement_Test.py [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+        self.__linked_list.add_first(3)
+        self.__linked_list.add_first(4)
+        self.__linked_list.add_first(5)
+        self.__linked_list.add_first(6)
+        self.__linked_list.add_first(7)
+        self.__linked_list.add_first(8)
+        self.__list_head = self.__linked_list.get_head()
+        
+    def testBaseCases(self):
+        self.__max_value = self.__linked_list.size()
```
Comment (flerdacodeu) [inline assignment1/KthToLastElement_Test.py]: You should either initialize `self.__max_value` in `setUp()` or you should use a local variable instead.
Note that, at the moment, if you run only the second test `testErrorCases` without running this test, the variable `self.__max_value` will not be initialized.

File: assignment1/KthToLastElement_Test.py [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+        for i in range(0, self.__max_value):
+            self.assertEqual(kth_to_last_element(self.__list_head, i), i)
+        
+    def testErrorCases(self):
+        with self.assertRaises(ValueError):
```
Comment (flerdacodeu) [inline assignment1/KthToLastElement_Test.py]: This is only testing the first operation, not both.
The code within the `with`-statement raises `ValueError`, but, at it does, the second statement is actually never executed.
You can do what you need here using two separate `with`-statements instead.

**Review by flerdacodeu — APPROVED** [review]
Just a couple of final comments.
Thanks for the changes!

File: assignment1/Exercise1.py (hunk 23 above)
Comment (etheodoridou) [reply assignment1/Exercise1.py]: Changed!

File: assignment1/Exercise1.py (hunk 24 above)
Comment (etheodoridou) [reply assignment1/Exercise1.py]: Changed it to raise a ValueError

File: assignment1/KthToLastElement_Test.py (hunk 25 above)
Comment (etheodoridou) [reply assignment1/KthToLastElement_Test.py]: Didn't notice that! Thanks for pointing it out. I have changed it now!

File: assignment1/KthToLastElement_Test.py (hunk 26 above)
Comment (etheodoridou) [reply assignment1/KthToLastElement_Test.py]: Done! Why is the assertRaises used with "with" as opposed to Java where it is just another assert statement?

Comment (flerdacodeu) [reply assignment1/KthToLastElement_Test.py]: The way the `with`-statement works is that it calls the `__exit__()` method on the object it created (the result of the call to `assertRaises()` in this case)  after executing the code within the `with`-statement, even if there is an exception. Basically it is similar to a `try/finally`-statement. This is used in this case to check that the assertion is actually raised. Compare with:

```python
  try:
    run()
    self.assertFalse("should not happen")
  except Exception, e:
    pass  # This is expected.
```

In Java, the way one can test for an exception can either be with a `try/finally`-statement (similar to the above); with the annotation `@Test(expected = XXX.class)`; or with a rule like `@Rule public final ExpectedException exception = ExpectedException.none();`.

Comment (etheodoridou) [reply assignment1/KthToLastElement_Test.py]: That makes sense, thank you for the explanation!
